File: autoit-mediapipe-com/src/binding/tasks/vision/face_detector.cc
#include "binding/tasks/vision/face_detector.h"

PTR_BRIDGE_IMPL(mediapipe::tasks::autoit::vision::face_detector::FaceDetectorResultRawCallback);

template<typename _Ty1, typename _Ty2>
inline const HRESULT autoit_to_callback(VARIANT const* const& in_val, _Ty1& out_val) {
	_Ty2 result_callback;
	HRESULT hr = autoit_to(in_val, result_callback);
	if (SUCCEEDED(hr)) {
		out_val = result_callback;
	}
	return hr;
}

const HRESULT autoit_to(VARIANT const* const& in_val, mediapipe::tasks::autoit::vision::face_detector::FaceDetectorResultCallback& out_val) {
	return autoit_to_callback<
		mediapipe::tasks::autoit::vision::face_detector::FaceDetectorResultCallback,
		mediapipe::tasks::autoit::vision::face_detector::FaceDetectorResultRawCallback
	>(in_val, out_val);
}

namespace {
	using namespace google::protobuf;
	using namespace mediapipe::autoit::packet_creator;
	using namespace mediapipe::autoit::packet_getter;
	using namespace mediapipe::tasks::autoit::components::utils;;
	using namespace mediapipe::tasks::autoit::core::base_options;
	using namespace mediapipe::tasks::autoit::core::task_info;
	using namespace mediapipe::tasks::autoit::vision::core::image_processing_options;
	using namespace mediapipe::tasks::autoit::vision::core::vision_task_running_mode;
	using namespace mediapipe::tasks::vision::face_detector::proto;
	using namespace mediapipe::tasks::autoit::vision::face_detector;

	using mediapipe::autoit::PacketsCallback;
	using mediapipe::tasks::core::PacketMap;

	using Detection = mediapipe::tasks::autoit::components::containers::detections::Detection;

	const std::string _DETECTIONS_OUT_STREAM_NAME = "detections";
	const std::string _DETECTIONS_TAG = "DETECTIONS";
	const std::string _NORM_RECT_STREAM_NAME = "norm_rect_in";
	const std::string _NORM_RECT_TAG = "NORM_RECT";
	const std::string _IMAGE_IN_STREAM_NAME = "image_in";
	const std::string _IMAGE_OUT_STREAM_NAME = "image_out";
	const std::string _IMAGE_TAG = "IMAGE";
	const std::string _TASK_GRAPH_NAME = "mediapipe.tasks.vision.face_detector.FaceDetectorGraph";
	const int64_t _MICRO_SECONDS_PER_MILLISECOND = 1000;

	[[nodiscard]] absl::StatusOr<std::shared_ptr<FaceDetectorResult>> _build_detection_result(const PacketMap& output_packets) {
		const auto& detector_out_packet = output_packets.at(_DETECTIONS_OUT_STREAM_NAME);
		if (detector_out_packet.IsEmpty()) {
			return std::make_shared<FaceDetectorResult>();
		}

		auto detection_result = std::make_shared<FaceDetectorResult>();

		MP_PACKET_ASSIGN_OR_RETURN(const auto& detection_proto_list, std::vector<mediapipe::Detection>, output_packets.at(_DETECTIONS_OUT_STREAM_NAME));
		for (const auto& detection : detection_proto_list) {
			detection_result->detections->push_back(Detection::create_from_pb2(detection));
		}

		return detection_result;
	}
}

namespace mediapipe::tasks::autoit::vision::face_detector {
	absl::StatusOr<std::shared_ptr<FaceDetectorGraphOptions>> FaceDetectorOptions::to_pb2() const {
		auto pb2_obj = std::make_shared<FaceDetectorGraphOptions>();

		if (base_options) {
			MP_ASSIGN_OR_RETURN(auto base_options_proto, base_options->to_pb2());
			pb2_obj->mutable_base_options()->CopyFrom(*base_options_proto);
		}
		pb2_obj->mutable_base_options()->set_use_stream_mode(running_mode != VisionTaskRunningMode::IMAGE);

		if (min_detection_confidence) pb2_obj->set_min_detection_confidence(*min_detection_confidence);
		if (min_suppression_threshold) pb2_obj->set_min_suppression_threshold(*min_suppression_threshold);

		return pb2_obj;
	}

	absl::StatusOr<std::shared_ptr<FaceDetector>> FaceDetector::create(
		const CalculatorGraphConfig& graph_config,
		VisionTaskRunningMode running_mode,
		mediapipe::autoit::PacketsCallback packet_callback
	) {
		using BaseVisionTaskApi = core::base_vision_task_api::BaseVisionTaskApi;
		return BaseVisionTaskApi::create(graph_config, running_mode, packet_callback, static_cast<FaceDetector*>(nullptr));
	}

	absl::StatusOr<std::shared_ptr<FaceDetector>> FaceDetector::create_from_model_path(const std::string& model_path) {
		auto base_options = std::make_shared<BaseOptions>(model_path);
		return create_from_options(std::make_shared<FaceDetectorOptions>(base_options, VisionTaskRunningMode::IMAGE));
	}

	absl::StatusOr<std::shared_ptr<FaceDetector>> FaceDetector::create_from_options(std::shared_ptr<FaceDetectorOptions> options) {
		PacketsCallback packet_callback = nullptr;

		if (options->result_callback) {
			packet_callback = [options](const PacketMap& output_packets) {
				const auto image_out_packet = output_packets.at(_IMAGE_OUT_STREAM_NAME);
				if (image_out_packet.IsEmpty()) {
					return;
				}

				MP_ASSIGN_OR_THROW(auto detection_result, _build_detection_result(output_packets)); // There is no other choice than throw in a callback to stop the execution
				MP_PACKET_ASSIGN_OR_THROW(const auto& image, Image, image_out_packet); // There is no other choice than throw in a callback to stop the execution
				auto timestamp_ms = output_packets.at(_DETECTIONS_OUT_STREAM_NAME).Timestamp().Value() / _MICRO_SECONDS_PER_MILLISECOND;

				options->result_callback(*detection_result, image, timestamp_ms);
			};
		}

		TaskInfo task_info;
		task_info.task_graph = _TASK_GRAPH_NAME;
		*task_info.input_streams = {
			_IMAGE_TAG + ":" + _IMAGE_IN_STREAM_NAME,
			_NORM_RECT_TAG + ":" + _NORM_RECT_STREAM_NAME
		};
		*task_info.output_streams = {
			_DETECTIONS_TAG + ":" + _DETECTIONS_OUT_STREAM_NAME,
			_IMAGE_TAG + ":" + _IMAGE_OUT_STREAM_NAME
		};
		MP_ASSIGN_OR_RETURN(task_info.task_options, options->to_pb2());

		MP_ASSIGN_OR_RETURN(auto config, task_info.generate_graph_config(options->running_mode == VisionTaskRunningMode::LIVE_STREAM));

		return create(
			*config,
			options->running_mode,
			std::move(packet_callback)
		);
	}

	absl::StatusOr<std::shared_ptr<FaceDetectorResult>> FaceDetector::detect(
		const Image& image,
		std::shared_ptr<ImageProcessingOptions> image_processing_options
	) {
		MP_ASSIGN_OR_RETURN(auto normalized_rect, convert_to_normalized_rect(image_processing_options, image, false));

		MP_ASSIGN_OR_RETURN(auto output_packets, _process_image_data({
			{ _IMAGE_IN_STREAM_NAME, std::move(*std::move(create_image(image))) },
			{ _NORM_RECT_STREAM_NAME, std::move(*std::move(create_proto(*normalized_rect.to_pb2()))) },
			}));

		return _build_detection_result(output_packets);
	}

	absl::StatusOr<std::shared_ptr<FaceDetectorResult>> FaceDetector::detect_for_video(
		const Image& image,
		int64_t timestamp_ms,
		std::shared_ptr<ImageProcessingOptions> image_processing_options
	) {
		MP_ASSIGN_OR_RETURN(auto normalized_rect, convert_to_normalized_rect(image_processing_options, image, false));

		MP_ASSIGN_OR_RETURN(auto output_packets, _process_video_data({
			{ _IMAGE_IN_STREAM_NAME, std::move(std::move(create_image(image))->At(
				Timestamp(timestamp_ms * _MICRO_SECONDS_PER_MILLISECOND)
			)) },
			{ _NORM_RECT_STREAM_NAME, std::move(std::move(create_proto(*normalized_rect.to_pb2()))->At(
				Timestamp(timestamp_ms * _MICRO_SECONDS_PER_MILLISECOND)
			)) },
			}));

		return _build_detection_result(output_packets);
	}

	absl::Status FaceDetector::detect_async(
		const Image& image,
		int64_t timestamp_ms,
		std::shared_ptr<ImageProcessingOptions> image_processing_options
	) {
		MP_ASSIGN_OR_RETURN(auto normalized_rect, convert_to_normalized_rect(image_processing_options, image, false));

		return _send_live_stream_data({
			{ _IMAGE_IN_STREAM_NAME, std::move(std::move(create_image(image))->At(
				Timestamp(timestamp_ms * _MICRO_SECONDS_PER_MILLISECOND)
			)) },
			{ _NORM_RECT_STREAM_NAME, std::move(std::move(create_proto(*normalized_rect.to_pb2()))->At(
				Timestamp(timestamp_ms * _MICRO_SECONDS_PER_MILLISECOND)
			)) },
			});
	}
}

File: src/declarations/tasks/vision/face_detector/face_detector.js
const ns = "mediapipe::tasks::vision::face_detector";
const nsProgId = ns.replaceAll("::", ".");
const fqn = `${ ns }::FaceDetector`;
const RunningMode = "tasks::vision::core::RunningMode";

module.exports = ({language}) => {
    const progid = `mediapipe.tasks.${ language }.vision.face_detector`;
    const utils = `mediapipe::tasks::${ language }::core::utils`;

    return [
        // expose face_detector properties like in mediapipe python
        [`${ progid }.`, "", ["/Properties"], [
            [`${ ns }::FaceDetectorOptions`, "FaceDetectorOptions", "", ["/R", "=this", "/S"]],
            [`${ ns }::FaceDetector`, "FaceDetector", "", ["/R", "=this", "/S"]],
            ["mediapipe::tasks::components::containers::DetectionResult", "FaceDetectorResult", "", ["/R", "=this", "/S"]],
        ], "", ""],

        [`struct ${ nsProgId }.FaceDetectorOptions`, "", [`/progid=${ progid }.FaceDetectorOptions`, "/Simple", "/DC"], [
            ["tasks::core::BaseOptions", "base_options", "", ["/RW"]],
            ["core::RunningMode", "running_mode", `${ RunningMode }::IMAGE`, ["/RW"]],
            ["float", "min_detection_confidence", "0.5f", ["/RW"]],
            ["float", "min_suppression_threshold", "0.3f", ["/RW"]],
            ["FaceDetectorResultCallback", "result_callback", "", [`/WExpr=${ utils }::CppConvertToResultCallback($value, $0, $hr)`]],
        ], "", ""],

        [`class ${ nsProgId }.FaceDetector`, "", [`/progid=${ progid }.FaceDetector`], [], "", ""],

        [`${ nsProgId }.FaceDetector.Create`, `std::shared_ptr<${ fqn }>`, ["=create_from_model_path", "/S", `/WrapAs=${ utils }::ReleaseToSharedPtr`], [
            ["std::string", "model_path", "", [`/Expr=std::unique_ptr<FaceDetectorOptions>(new FaceDetectorOptions{ .base_options{ .model_asset_path{ $0 } }, .running_mode{ ${ RunningMode }::IMAGE } })`]],
        ], "", ""],

        [`${ nsProgId }.FaceDetector.Create`, `std::shared_ptr<${ fqn }>`, ["=create_from_options", "/S", `/WrapAs=${ utils }::ReleaseToSharedPtr`], [
            ["std::shared_ptr<FaceDetectorOptions>", "options", "", ["/Expr=tasks::autoit::core::utils::ConvertToUniquePtr($0)"]],
        ], "", ""],

        [`${ nsProgId }.FaceDetector.Detect`, "absl::StatusOr<FaceDetectorResult>", ["=detect"], [
            ["Image", "image", "", []],
            ["std::optional<core::ImageProcessingOptions>", "image_processing_options", "std::nullopt", []],
        ], "", ""],

        [`${ nsProgId }.FaceDetector.DetectForVideo`, "absl::StatusOr<FaceDetectorResult>", ["=detect_for_video"], [
            ["Image", "image", "", []],
            ["uint64_t", "timestamp_ms", "", []],
            ["std::optional<core::ImageProcessingOptions>", "image_processing_options", "std::nullopt", []],
        ], "", ""],

        [`${ nsProgId }.FaceDetector.DetectAsync`, "absl::Status", ["=detect_async"], [
            ["Image", "image", "", []],
            ["uint64_t", "timestamp_ms", "", []],
            ["std::optional<core::ImageProcessingOptions>", "image_processing_options", "std::nullopt", []],
        ], "", ""],

        [`${ nsProgId }.FaceDetector.Close`, "absl::Status", ["=close"], [], "", ""],
    ];
};
